tests/testsuite/publish.rs

```diff
@@ -1800,3 +1800,228 @@ See [..]
 
     validate_upload_bar();
 }
+
+#[cargo_test]
+fn with_readme() {
+    registry::init();
+
+    let p = git::new("foo", |p| {
+        p.file(
+            "Cargo.toml",
+            r#"
+                [project]
+                name = "foo"
+                version = "0.0.1"
+                authors = []
+                license = "MIT"
+                description = "foo"
+                readme = "README"
+            "#,
+        )
+        .file("src/main.rs", "fn main() {}")
+        .file("README", "HELLO")
+    });
+
+    p.cargo("publish --no-verify --token sekrit").run();
+
+    publish::validate_upload(
+        r#"
+        {
+          "authors": [],
+          "badges": {},
+          "categories": [],
+          "deps": [],
+          "description": "foo",
+          "documentation": null,
+          "features": {},
+          "homepage": null,
+          "keywords": [],
+          "license": "MIT",
+          "license_file": null,
+          "links": null,
+          "name": "foo",
+          "readme": "HELLO",
+          "readme_file": "README",
+          "repository": null,
+          "vers": "0.0.1"
+          }
+        "#,
+        "foo-0.0.1.crate",
+        &[
+            "Cargo.lock",
+            "Cargo.toml",
+            "Cargo.toml.orig",
+            "src/main.rs",
+            "README",
+            ".cargo_vcs_info.json",
+        ],
+    );
+}
+
+#[cargo_test]
+fn with_non_root_readme() {
+    registry::init();
+
+    let p = git::new("foo", |p| {
+        p.file(
+            "Cargo.toml",
+            r#"
+                [project]
+                name = "foo"
+                version = "0.0.1"
+                authors = []
+                license = "MIT"
+                description = "foo"
+                readme = "docs/README"
+            "#,
+        )
+        .file("src/main.rs", "fn main() {}")
+        .file("docs/README", "HELLO")
+    });
+
+    p.cargo("publish --no-verify --token sekrit").run();
+
+    publish::validate_upload(
+        r#"
+        {
+          "authors": [],
+          "badges": {},
+          "categories": [],
+          "deps": [],
+          "description": "foo",
+          "documentation": null,
+          "features": {},
+          "homepage": null,
+          "keywords": [],
+          "license": "MIT",
+          "license_file": null,
+          "links": null,
+          "name": "foo",
+          "readme": "HELLO",
+          "readme_file": "docs/README",
```

Comment:
this will probably end up failing on windows with a backslash here? I think we'd want to send forwardslashes to crates.io here. How to convert?

Reply:
Cool. I was unable to get the tests to actually cover this because they use weak json matching (that converts \ to /). I made the `replace` change and manually confirmed it works with some println's in the `compare.rs` to make sure.

I wonder if there's somewhere we can put some assert in the test implementation of the registry. I was having trouble tracking down what `cargo publish` does in the unit tests - maybe the test backend can assert for fwd slashes only? Do you know where to look for that?